### PR TITLE
Add draft component architecture doc

### PR DESCRIPTION
**Pull request by cpuguy83**
Adds basic architecture details for componentization work, and some
starter protobuf definitions to work from.

This doc is still in draft form and should be considered as such even
after merge until the document states otherwise.
The intent of this commit is not to put something in the repo that is
gospel for the componentization work, but rather get the work kicked off
so a larger group can start taking a look, requesting changes, etc.

Basically, at this point this document should change or be validated by
implementation (though realistically, the document is not complete).
Over time there should be fewer changes and this would be the source of
truth for implementing a moby component.

ping @moby/moby-maintainers

### CODE REVIEW

File: components/architecture.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+- moby-volume-store - Provides basic CRUD operations for volumes
+	- Dependencies:
+		- moby-plugin-store
+- moby-volume-runtime - Sets up volumes for a container
+	- Dependencies:
+		- moby-volume-store
+
+#### Image Management
+
+- moby-layer-store - Storage for image layers
```
Comment (dmcgowan) [inline components/architecture.md]: Is there a motivation for having this be a separate component? We want this part of the code to be deprecated by containerd and only have containerd implement the image store interface for doing image management. The implementations themselves will already have sub-components but does that make it a "moby component"?

Comment (cpuguy83) [reply components/architecture.md]: So this wouldn't be the literal layer store that's currently in moby/moby.

This may not be a component by itself, but rather part of one (a component may comprise of more than 1 service). But it's something we need to implement.
What will actually be in here I'm not sure yet, but we need to support the existing drivers (even if re-written to be snapshotters) at least.

File: components/architecture.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+A component:
+- MUST provide a GRPC endpoint with the following services on a node-local socket (such as a unix socket or named pipe):
+	- MUST provide [GRPC healthcheck service](https://github.com/grpc/grpc/blob/master/doc/health-checking.md)
+	- MUST provide [Version service](api/services/component/version/v1)
+	- SHOULD provide [Metrics service](api/services/component/metrics/v1)
+- MAY expose a separate API for external consumption
+- SHOULD use GRPC to publish component-level services
+- SHOULD be easily deployed as a container (the primary packaging format for
+Moby Core, see [Packaging](#Packaging)).
```
Comment (aaronlehmann) [inline components/architecture.md]: How many processes do you expect will be involved in providing functionality equivalent to Docker?

If components are separate Go binaries running in their own containers, they will be relatively heavyweight, in terms of size and memory footprint. If covering the Docker feature set involves, say, 10 binaries of 30 MB each, that may be a nonstarter for certain embedded applications. I could also see it being an issue for memory-limited VMs, especially in the cloud. If we are inefficient with resources it could push people away from containerization, or towards more bare-bones solutions like standalone containerd.

Comment (cpuguy83) [reply components/architecture.md]: I agree this is something we'll need to watch out for.
However I would prefer to optimize for this later and I don't think it's strictly an architecture problem.

dockerd itself is only 33M, so I expect components to come in quite a bit smaller. I know GRPC brings a hefty price tag in that regard (around 10MB last I saw), though.

Comment (stevvooe) [reply components/architecture.md]: > GRPC brings a hefty price tag in that regard (around 10MB last I saw)

Wow. We should file that as a bug. I noticed that we are taking on way too many dependencies to carry pre-http2 Go versions. I'll bring that up at the community meeting.

File: components/api/services/component/metrics/v1/metrics.proto [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+package moby.core.services.metrics.v1;
+
+option go_package = "metrics";
+
+import "github.com/prometheus/client_model/metrics.proto";
+
+// Metrics provides support for publishing prometheus style metrics on a GRPC endpoint.
+// This would be used by a client to collect metrics and expose to an external service.
+service Metrics {
+	rpc CollectMetrics(CollectMetricsRequest) returns (CollectMetricsResponse);
```
Comment (stevvooe) [inline components/api/services/component/metrics/v1/metrics.proto]: @crosbymichael PTAL

File: components/api/services/containers/rootfs/v1/rootfs.proto [hunk 4]
```diff
@@ -0,0 +1,49 @@
+syntax = "proto3";
+
+package moby.core.services.containers.rootfs.v1;
+
+option go_package = "rootfs";
+
+import "github.com/moby/moby/components/api/types/container/container.proto";
+import "github.com/moby/moby/components/api/types/file.proto";
+
+service RootfsService {
```
Comment (AkihiroSuda) [inline components/api/services/containers/rootfs/v1/rootfs.proto]: Can we just drop this service and expose bare containerd snapshotter service?
https://github.com/containerd/containerd/blob/master/api/services/snapshot/v1/snapshots.proto

That would deduplicate much effort for image&snapshot&rootfs abstraction.

cc @stevvooe @dmcgowan

File: components/api/services/containers/runtime/v1/runtime.proto [hunk 5]
```diff
@@ -0,0 +1,43 @@
+syntax = "proto3";
+
+package moby.core.services.containers.runtime.v1;
+
+option go_package = "runtime";
+
+import "github.com/moby/moby/components/api/types/container/container.proto";
+import "google/protobuf/duration.proto";
+
+service ContainerRuntimeService {
```
Comment (AkihiroSuda) [inline components/api/services/containers/runtime/v1/runtime.proto]: Similarly, can we just expose (almost) bare containerd service? (with some `labels` and `extensions`)
https://github.com/containerd/containerd/blob/master/api/services/containers/v1/containers.proto

File: components/api/services/containers/rootfs/v1/rootfs.proto (hunk 4 above)
Comment (cpuguy83) [reply components/api/services/containers/rootfs/v1/rootfs.proto]: Mostly this is for compatibility for abstracting docker v1 compatibility into a nice service that's not dependent on the backend implementation.
We could expose the bare containerd service and just say "moby is built on containerd end of story".

**Comment by thaJeztah** [comment]
Let's close this for now, but we can continue the discussion

/cc @crosbymichael @dhiltgen
